prot.c

```diff
@@ -1,15 +1,29 @@
+#include "prot.h"
 #include <sys/types.h>
-#include <unistd.h>
 #include <grp.h>
-#include "prot.h"
+#include <unistd.h>
 
-int prot_gid(gid_t gid)
+#ifndef GROUP_COUNT
+/* use a sane default, but one can override it from outside if there are
+ * higher requirements. */
+#define GROUP_COUNT 8
+#endif
+int prot_gids(const char *user, gid_t gid)
 {
-  if (setgroups(1,&gid) == -1) return -1;
-  return setgid(gid); /* _should_ be redundant, but on some systems it isn't */
-}
+  gid_t gids[GROUP_COUNT];
+  int gcount = GROUP_COUNT;
 
-int prot_uid(uid_t uid)
-{
-  return setuid(uid);
+  if (user) {
+    int r = getgrouplist(user, gid, gids, &gcount);
```

Comment:
We could just use `initgroups` no?

Reply:
Please verify getgrouplist works on MacOS, as getgrouplist uses int* in the prototype there. But sizeof gid_t is 4, so it should work. I still recommend to use initgroups tho.

Reply:
I have updated the code, but have not tested it yet.